timing/testmode.go

```diff
@@ -101,38 +125,74 @@ func (s *Scheduler) setNextTrigger(when time.Time) *Scheduler {
 		triggers = append(triggers, trigger{s, when})
 	}
 	sort.Sort(triggers)
-	return s
-}
-
-func (s *Scheduler) nextRepeatingTick() time.Time {
-	elapsedIntervals := Now().Sub(s.startTime) / s.interval
-	return s.startTime.Add(s.interval * (elapsedIntervals + 1))
 }
 
-func (s *Scheduler) testModeAt(when time.Time) *Scheduler {
+// At implements the schedulerImpl interface.
+func (s *testModeScheduler) At(when time.Time, f func()) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	l.Fine("%s At[Test](%v)", l.ID(s), when)
-	return s.setNextTrigger(when)
+	s.f = f
+	s.setNextTrigger(when)
 }
 
-func (s *Scheduler) testModeAfter(delay time.Duration) *Scheduler {
+// After implements the schedulerImpl interface.
+func (s *testModeScheduler) After(delay time.Duration, f func()) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	l.Fine("%s After[Test](%v)", l.ID(s), delay)
-	return s.setNextTrigger(Now().Add(delay))
+	s.f = f
+	s.setNextTrigger(Now().Add(delay))
 }
 
-func (s *Scheduler) testModeEvery(interval time.Duration) *Scheduler {
+// Every implements the schedulerImpl interface.
+func (s *testModeScheduler) Every(interval time.Duration, f func()) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	l.Fine("%s Every[Test](%v)", l.ID(s), interval)
+	s.interval = interval
+	s.alignOffset = Now().Sub(Now().Truncate(interval))
+	s.f = f
+	s.setNextTrigger(s.nextRepeatingTick())
+}
+
+// EveryAlign implements the schedulerImpl interface.
+func (s *testModeScheduler) EveryAlign(interval time.Duration, offset time.Duration, f func()) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.startTime = Now()
+	l.Fine("%s EveryAlign[Test](%v)", l.ID(s), interval)
 	s.interval = interval
-	return s.setNextTrigger(s.nextRepeatingTick())
+	s.alignOffset = offset
+	s.f = f
+	s.setNextTrigger(s.nextRepeatingTick())
 }
 
-func (s *Scheduler) testModeStop() {
+// Stop implements the schedulerImpl interface.
+func (s *testModeScheduler) Stop() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	l.Fine("%s Stop[Test]", l.ID(s))
+	s.f = nil
 	s.setNextTrigger(time.Time{})
 }
 
+// Close implements the schedulerImpl interface.
+func (s *testModeScheduler) Close() {
+	s.Stop()
+}
+
+func (s *testModeScheduler) nextRepeatingTick() time.Time {
+	now := Now()
+	next := now.Truncate(s.interval).Add(s.alignOffset)
+	if !next.After(now) {
+		next = next.Add(s.interval)
+	}
+	if !next.After(now) {
+		panic("RealtimeScheduler.nextRepeatingTick: bug: !next.After(now)")
```

Comment:
Did you add this because of a confusing failure during your testing? If not, I don't think we need to be extra defensive here.

Reply:
At some point I did have `next.Add(s.interval)` instead of `next = next.Add(s.interval)`. Extra sanity assertions make code more believable, and if there will be another bug somewhere nearby, they will allow to skip this fragment instead of verifying it again.

But I honestly don't have a particularly strong opinion about this check, so I'm fine with removing it.

Reply:
Ah. Probably better to cover something like that in a test (but then again, this is testmode, so maybe either place is fine). If you leave it in, update the panic message to say "TestModeScheduler" or something instead of "RealtimeScheduler".

Reply:
There're a couple of places with the same piece of code. I think either all or none of them should have this sanity assertion panic.

Reply:
> Probably better to cover something like that in a test 

This `next = next.Add(interval)` is covered by the tests simply because it happens most of the time, though. And if it was wrong, timer would fire immediately, which is also covered by the tests.

Reply:
Given that this fragment appears in three different places, I did the obvious thing of moving it to a separate function.